File: php/src/Core/Json.php
<?php

declare(strict_types=1);

namespace SolanaMpp\Core;

use InvalidArgumentException;

/**
 * Narrows decoded JSON values before protocol parsing.
 */
final class Json
{
    /**
     * Encode a value as RFC 8785 canonical JSON.
     *
     * Sorts object keys by UTF-16 code-unit order, serializes numbers per ES6 ToString
     * (ECMA-262 7.1.12.1), and rejects NaN, Infinity, and lone surrogates.
     */
    public static function canonicalize(mixed $value): string
    {
        return self::encodeValue($value);
    }

    private static function encodeValue(mixed $value): string
    {
        if ($value === null) {
            return 'null';
        }
        if (is_bool($value)) {
            return $value ? 'true' : 'false';
        }
        if (is_int($value)) {
            return (string)$value;
        }
        if (is_float($value)) {
            return self::encodeNumber($value);
        }
        if (is_string($value)) {
            return self::encodeString($value);
        }
        if (is_array($value)) {
            if (array_is_list($value)) {
                $parts = array_map(fn ($item) => self::encodeValue($item), $value);
                return '[' . implode(',', $parts) . ']';
            }
            return self::encodeObject($value);
        }
        throw new InvalidArgumentException('unsupported JSON value');
    }

    /**
     * @param array<array-key, mixed> $value
     */
    private static function encodeObject(array $value): string
    {
        $stringKeyed = [];
        foreach ($value as $key => $item) {
            $stringKey = (string)$key;
            if (array_key_exists($stringKey, $stringKeyed)) {
                throw new InvalidArgumentException('duplicate object key');
            }
            $stringKeyed[$stringKey] = $item;
        }
        $keys = array_keys($stringKeyed);
        usort($keys, static function (string $a, string $b): int {
            return self::compareUtf16($a, $b);
        });
        $parts = [];
        foreach ($keys as $key) {
            $parts[] = self::encodeString($key) . ':' . self::encodeValue($stringKeyed[$key]);
        }
        return '{' . implode(',', $parts) . '}';
    }

    /**
     * Compare two UTF-8 strings by UTF-16 code-unit order (RFC 8785 sec 3.2.3).
     */
    private static function compareUtf16(string $a, string $b): int
    {
        $au = mb_convert_encoding($a, 'UTF-16BE', 'UTF-8');
        $bu = mb_convert_encoding($b, 'UTF-16BE', 'UTF-8');
        $aLen = strlen($au);
        $bLen = strlen($bu);
        $n = min($aLen, $bLen);
        for ($i = 0; $i < $n; $i += 2) {
            $ax = (ord($au[$i]) << 8) | ord($au[$i + 1]);
            $bx = (ord($bu[$i]) << 8) | ord($bu[$i + 1]);
            if ($ax !== $bx) {
                return $ax <=> $bx;
            }
        }
        return $aLen <=> $bLen;
    }

    /**
     * ES6 ToString (ECMA-262 7.1.12.1) number serialization for JCS (RFC 8785 sec 3.2.2.3).
     *
     * Plain decimal notation when the shortest round-trip representation has decimal exponent k
     * with -6 < k <= 20, exponential form otherwise.
     */
    private static function encodeNumber(float $value): string
    {
        if (is_nan($value)) {
            throw new InvalidArgumentException('cannot encode NaN');
        }
        if (is_infinite($value)) {
            throw new InvalidArgumentException('cannot encode Infinity');
        }
        if ($value === 0.0) {
            return '0';
        }
        $sign = $value < 0 ? '-' : '';
        [$digits, $k] = self::shortestDigitsAndExponent(abs($value));
        return self::formatEs6Number($sign, $digits, $k);
    }

    /**
     * @return array{0: string, 1: int}
     */
    private static function shortestDigitsAndExponent(float $absValue): array
    {
        // PHP's (string) cast respects serialize_precision but is not always shortest round-trip
        // for edge values like 0.30000000000000004. Use sprintf with the explicit shortest-trip
        // %.17g and the shorter %.15g fallback if it round-trips.
        $short = sprintf('%.15g', $absValue);
        if ((float)$short === $absValue) {
            $repr = $short;
        } else {
            $repr = sprintf('%.17g', $absValue);
        }
        if (stripos($repr, 'e') !== false) {
            $parts = preg_split('/[eE]/', $repr);
            if (!is_array($parts) || count($parts) !== 2) {
                return [$repr, 0];
            }
            $mantissa = $parts[0];
            $expInt = (int)$parts[1];
        } else {
            $mantissa = $repr;
            $expInt = 0;
        }
        $dotPos = strpos($mantissa, '.');
        if ($dotPos === false) {
            $intPart = $mantissa;
            $fracPart = '';
        } else {
            $intPart = substr($mantissa, 0, $dotPos);
            $fracPart = substr($mantissa, $dotPos + 1);
        }
        $combined = $intPart . $fracPart;
        $stripped = ltrim($combined, '0');
        $leadingZeros = strlen($combined) - strlen($stripped);
        $digits = rtrim($stripped, '0');
        if ($digits === '') {
            $digits = '0';
        }
        $decimalExponent = $expInt + strlen($intPart) - 1 - $leadingZeros;
        return [$digits, $decimalExponent];
    }

    private static function formatEs6Number(string $sign, string $digits, int $k): string
    {
        $n = strlen($digits);
        if ($k >= 0 && $k <= 20) {
            if ($n <= $k + 1) {
                return $sign . $digits . str_repeat('0', $k + 1 - $n);
            }
            return $sign . substr($digits, 0, $k + 1) . '.' . substr($digits, $k + 1);
        }
        if ($k < 0 && $k > -7) {
            return $sign . '0.' . str_repeat('0', -$k - 1) . $digits;
        }
        $mantissa = $n === 1 ? $digits : ($digits[0] . '.' . substr($digits, 1));
        $expSign = $k >= 0 ? '+' : '-';
        return $sign . $mantissa . 'e' . $expSign . abs($k);
    }

    /**
     * Emit a JCS-conformant JSON string literal (RFC 8785 sec 3.2.2.2), rejecting lone surrogates.
     */
    private static function encodeString(string $value): string
    {
        // Validate UTF-8 then walk codepoints.
        if (!mb_check_encoding($value, 'UTF-8')) {
            throw new InvalidArgumentException('invalid UTF-8 in string');
        }
        $codepoints = mb_str_split($value, 1, 'UTF-8');
        $buf = '"';
        foreach ($codepoints as $char) {
            $cp = mb_ord($char, 'UTF-8');
            if ($cp >= 0xD800 && $cp <= 0xDFFF) {
                throw new InvalidArgumentException('lone surrogate in string');
            }
            $buf .= match (true) {
                $char === '\\' => '\\\\',
                $char === '"' => '\\"',
                $char === "\b" => '\\b',
                $char === "\t" => '\\t',
                $char === "\n" => '\\n',
                $char === "\f" => '\\f',
                $char === "\r" => '\\r',
                $cp < 0x20 => sprintf('\\u%04x', $cp),
                default => $char,
            };
        }
        return $buf . '"';
    }

    /**
     * Require a decoded JSON value to be an object-shaped array.
     *
     * @return array<string, mixed>
     */
    public static function object(mixed $value, string $field): array
    {
        if (!is_array($value)) {
            throw new InvalidArgumentException($field . ' must be an object');
        }

        $object = [];
        foreach ($value as $key => $item) {
            if (!is_string($key)) {
                throw new InvalidArgumentException($field . ' must be an object');
            }
            $object[$key] = $item;
        }

        return $object;
    }

    /**
     * Require a decoded JSON value to be a string.
     */
    public static function string(mixed $value, string $field): string
    {
        if (!is_string($value)) {
            throw new InvalidArgumentException($field . ' must be a string');
        }

        return $value;
    }

    /**
     * Return a decoded JSON string or a default when the field is absent.
     */
    public static function optionalString(mixed $value, string $field, string $default = ''): string
    {
        if ($value === null) {
            return $default;
        }

        return self::string($value, $field);
    }

    /**
     * Return a decoded JSON integer or null when the field is absent.
     */
    public static function optionalInt(mixed $value, string $field): ?int
    {
        if ($value === null) {
            return null;
        }
        if (!is_int($value)) {
            throw new InvalidArgumentException($field . ' must be an integer');
        }

        return $value;
    }
}

File: ruby/lib/mpp/core/json.rb
# frozen_string_literal: true

require "json"

module Mpp
  module Core
    # RFC 8785 canonical JSON encoder for MPP header payloads.
    #
    # Vendors a small JCS implementation rather than delegating to JSON.generate so the
    # ordering, number serialization, and surrogate validation rules match the Rust spine.
    # See RFC 8785 sec 3.2.2 and sec 3.2.3.
    #
    # @see https://datatracker.ietf.org/doc/html/rfc8785 RFC 8785 JSON Canonicalization Scheme
    # @see https://tc39.es/ecma262/multipage/abstract-operations.html#sec-numeric-types-number-tostring
    #      ECMA-262 Number::toString algorithm
    module Json
      module_function

      # Encode a Ruby object with stable object key ordering (UTF-16 code-unit).
      def canonical_generate(value)
        encode_value(value)
      end

      # Decode JSON and preserve object keys as strings.
      def parse(value)
        JSON.parse(value)
      rescue JSON::ParserError => error
        raise ArgumentError, "invalid JSON: #{error.message}"
      end

      # ── private encoders ──

      class << self
        private

        def encode_value(value)
          case value
          when Hash then encode_object(value)
          when Array then "[" + value.map { |item| encode_value(item) }.join(",") + "]"
          when String then encode_string(value)
          when Integer then value.to_s
          when Float then encode_number(value)
          when true then "true"
          when false then "false"
          when nil then "null"
          else
            raise ArgumentError, "unsupported JSON value #{value.class}"
          end
        end

        def encode_object(hash)
          string_keys = hash.each_with_object({}) do |(key, val), memo|
            string_key = key.is_a?(Symbol) ? key.to_s : key
            raise ArgumentError, "object key must be a string" unless string_key.is_a?(String)
            raise ArgumentError, "duplicate object key #{string_key.inspect}" if memo.key?(string_key)

            memo[string_key] = val
          end
          ordered = string_keys.keys.sort_by { |k| utf16_code_units(k) }
          parts = ordered.map { |k| encode_string(k) + ":" + encode_value(string_keys.fetch(k)) }
          "{" + parts.join(",") + "}"
        end

        # Convert a UTF-8 string into an array of UTF-16 code units for ordering (RFC 8785 sec 3.2.3).
        def utf16_code_units(string)
          # encode! through UTF-16BE then split into 16-bit units; sort_by uses array comparison.
          utf16 = string.encode("UTF-16BE", invalid: :replace, undef: :replace).bytes
          units = []
          i = 0
          while i < utf16.length
            units << ((utf16[i] << 8) | utf16[i + 1])
            i += 2
          end
          units
        end

        # ES6 ToString (ECMA-262 7.1.12.1) number serialization for JCS (RFC 8785 sec 3.2.2.3).
        #
        # Mirrors V8/JavaScriptCore semantics: plain decimal notation when the shortest
        # round-trip representation has decimal exponent k with -6 < k <= 20, exponential
        # form ("Ne+EE") otherwise.
        def encode_number(value)
          raise ArgumentError, "cannot encode NaN" if value.nan?
          raise ArgumentError, "cannot encode Infinity" if value.infinite?
          return "0" if value.zero? # collapses -0 to "0"

          sign = value.negative? ? "-" : ""
          digits, k = shortest_digits_and_exponent(value.abs)
          format_es6_number(sign, digits, k)
        end

        # Return [digits, k] where digits is the shortest decimal mantissa and k is the
        # decimal exponent of the leading digit, so that value = 0.<digits> * 10^(k+1).
        def shortest_digits_and_exponent(abs_value)
          repr = abs_value.to_s # Ruby Float#to_s is shortest-round-trip.
          if repr.include?("e")
            mantissa, exp_str = repr.split("e")
            exp_int = exp_str.to_i
          else
            mantissa = repr
            exp_int = 0
          end
          int_part, frac_part = mantissa.split(".")
          frac_part ||= ""
          combined = int_part + frac_part
          # k_repr: the exponent of the leading digit if we treat 'combined' as 0.<combined> * 10^(int_part.length + exp_int).
          # i.e. value = combined * 10^(exp_int - frac_part.length).
          # decimal_exponent_of_leading_nonzero = (exp_int + int_part.length) - (number of leading zeros stripped) - 1.
          stripped = combined.sub(/\A0+/, "")
          leading_zeros = combined.length - stripped.length
          digits = stripped.sub(/0+\z/, "")
          digits = "0" if digits.empty?
          decimal_exponent = exp_int + int_part.length - 1 - leading_zeros
          [digits, decimal_exponent]
        end

        # Render digits + decimal exponent k as ES6 ToString.
        # Uses plain decimal when -6 < k <= 20, otherwise exponential.
        def format_es6_number(sign, digits, k)
          n = digits.length
          if k.between?(0, 20)
            if n <= k + 1
              return sign + digits + ("0" * (k + 1 - n))
            end
            return sign + digits[0, k + 1] + "." + digits[(k + 1)..]
          end
          if k < 0 && k > -7
            return sign + "0." + ("0" * (-k - 1)) + digits
          end
          mantissa = (n == 1) ? digits : (digits[0] + "." + digits[1..])
          exp_sign = (k >= 0) ? "+" : "-"
          sign + mantissa + "e" + exp_sign + k.abs.to_s
        end

        ESCAPE_TABLE = {
          "\b" => "\\b",
          "\t" => "\\t",
          "\n" => "\\n",
          "\f" => "\\f",
          "\r" => "\\r",
          "\"" => "\\\"",
          "\\" => "\\\\"
        }.freeze

        # Emit a JCS-conformant JSON string literal (RFC 8785 sec 3.2.2.2), rejecting lone surrogates.
        def encode_string(string)
          raise ArgumentError, "object key must be a string" unless string.is_a?(String)

          # Validate UTF-8 and reject any string containing a lone surrogate codepoint.
          codepoints = string.encode(Encoding::UTF_8).codepoints
          codepoints.each do |cp|
            raise ArgumentError, "lone surrogate in string" if cp.between?(0xD800, 0xDFFF)
          end

          buf = +"\""
          codepoints.each do |cp|
            buf << if (esc = ESCAPE_TABLE[[cp].pack("U")])
              esc
            elsif cp < 0x20
              format("\\u%04x", cp)
            elsif cp <= 0x7E
              cp.chr(Encoding::UTF_8)
            else
              # Non-ASCII: emit raw UTF-8 (JCS does not normalize, RFC 8785 sec 3.2.4).
              [cp].pack("U")
            end
          end
          buf << "\""
          buf
        end
      end
    end
  end
end
